### PR TITLE
SONAR-10181 mandatory WS Params are automatically validated on retrieval

### PR DESCRIPTION
**Pull request by gjambet**

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java [hunk 1]
```diff
@@ -202,6 +207,13 @@ private static void validateMaximumValue(String key, WebService.Param definition
     checkArgument(valueAsInt <= maximumValue, "'%s' value (%s) must be less than %s", key, valueAsInt, maximumValue);
   }
 
+  private void validateRequiredValue(String key, WebService.Param definition, String value) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: MAJOR Make "validateRequiredValue" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java [hunk 2]
```diff
@@ -27,6 +27,7 @@
 import java.util.stream.Collectors;
 import javax.annotation.CheckForNull;
 import javax.annotation.Nullable;
+import org.apache.commons.lang.StringUtils;
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: MINOR Remove this unused import 'org.apache.commons.lang.StringUtils'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java [hunk 3]
```diff
@@ -69,7 +69,11 @@ public void setLocalConnector(LocalConnector lc) {
   @CheckForNull
   public String param(String key) {
     WebService.Param definition = action.param(key);
-    String valueOrDefault = defaultString(readParam(key, definition), definition.defaultValue());
+
+    String rawValue = readParam(key, definition);
+    validateRequiredValue(key, definition, rawValue);
```
Comment (teryk) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: just before line 78

File: sonar-plugin-api/src/test/java/org/sonar/api/server/ws/internal/ValidatingRequestTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.sonar.api.server.ws.WebService;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.CALLS_REAL_METHODS;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ValidatingRequestTest {
```
Comment (teryk) [inline sonar-plugin-api/src/test/java/org/sonar/api/server/ws/internal/ValidatingRequestTest.java]: drop and test in org.sonar.server.ws.WebServiceEngineTest

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/template/TemplateGroupsAction.java [hunk 5]
```diff
@@ -83,7 +85,10 @@ public void define(WebService.NewController context) {
         "When this parameter is not set, only group having at least one permission are returned.")
       .setExampleValue("eri");
 
-    createProjectPermissionParameter(action);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/template/TemplateGroupsAction.java]: suggestion: keep the method and add isRequired(false)

Comment (gjambet) [reply server/sonar-server/src/main/java/org/sonar/server/permission/ws/template/TemplateGroupsAction.java]: done, method restored w a parameter for mandatory

**Pull request merged**